packages/replayio/src/utils/protocol/ProtocolClient.ts

```diff
@@ -0,0 +1,123 @@
+import assert from "assert";
+import { Agent } from "http";
+import WebSocket from "ws";
+import { replayServer } from "../../config";
+import { getAccessToken } from "../authentication/getAccessToken";
+import { Deferred, STATUS_PENDING, createDeferred } from "../createDeferred";
+import { ProtocolError } from "./ProtocolError";
+import { setAccessToken } from "./api/setAccessToken";
+import { debug } from "./debug";
+
+export default class ProtocolClient {
+  private deferredAuthenticated = createDeferred<boolean>();
+  private eventListeners = new Map();
+  private nextMessageId = 1;
+  private pendingMessages: Map<number, Deferred<any>> = new Map();
+  private socket: WebSocket;
+
+  constructor({ agent }: { agent?: Agent } = {}) {
+    debug("Creating WebSocket for %s with %o", replayServer, { agent });
+
+    this.socket = new WebSocket(replayServer, {
```

Comment:
it would be great to make this websocket reconnectable - it's somewhat easy right now to succeed with an upload through retries just to fail on an already failed WS connection

Reply:
Sure, but going to defer this for now -> PRO-82